quickwit/quickwit-common/src/tower/rate_estimator.rs

```diff
@@ -17,10 +17,236 @@
 // You should have received a copy of the GNU Affero General Public License
 // along with this program. If not, see <http://www.gnu.org/licenses/>.
 
-use std::time::Instant;
+use std::num::NonZeroUsize;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::Arc;
+use std::time::{Duration, Instant};
 
-use super::Rate;
+use super::{ConstantRate, Rate};
 
 pub trait RateEstimator: Rate {
     fn update(&mut self, started_at: Instant, ended_at: Instant, work: u64);
 }
+
+/// Simple moving average rate estimator. Tracks the average rate of work over a sliding time
+/// window.
+#[derive(Debug, Clone)]
+pub struct SmaRateEstimator {
+    inner: Arc<InnerSmaRateEstimator>,
+}
+
+#[derive(Debug)]
+struct InnerSmaRateEstimator {
+    anchor: Instant,
+    buckets: Box<[Bucket]>,
+    bucket_period_secs: u64,
+    bucket_period_millis: u64,
+    num_buckets: u64,
+    period: Duration,
+    period_millis: u64,
+}
+
+impl SmaRateEstimator {
+    /// Creates a new simple moving average rate estimator.
+    ///
+    /// # Panics
+    ///
+    /// This function panics if `bucket_period` is < 1s  or `period` is < 1ms.
+    pub fn new(num_buckets: NonZeroUsize, bucket_period: Duration, period: Duration) -> Self {
+        assert!(bucket_period.as_secs() > 0);
+        assert!(period.as_millis() > 0);
+
+        let mut buckets = Vec::with_capacity(num_buckets.get());
+        for bucket_ord in 0..num_buckets.get() {
+            buckets.push(Bucket::new(bucket_ord as u64, 0));
+        }
+        let inner = InnerSmaRateEstimator {
+            anchor: Instant::now(),
+            buckets: buckets.into_boxed_slice(),
+            bucket_period_secs: bucket_period.as_secs(),
+            bucket_period_millis: bucket_period.as_millis() as u64,
+            num_buckets: num_buckets.get() as u64,
+            period,
+            period_millis: period.as_millis() as u64,
+        };
+        Self {
+            inner: Arc::new(inner),
+        }
+    }
+
+    /// Seeds the rate estimator with an initial rate.
+    pub fn with_initial_rate(self, initial_rate: ConstantRate) -> Self {
+        let initial_work = initial_rate.work() * self.inner.bucket_period_millis
+            / initial_rate.period().as_millis() as u64;
+        for i in 0..self.inner.num_buckets {
+            self.inner.buckets[i as usize].update(initial_work);
+        }
+        self
+    }
+}
+
+impl Rate for SmaRateEstimator {
+    fn work(&self) -> u64 {
```

Comment:
// Returns an estimate of the amount of work to be achieved in a given `period`.